common/lib/common-utils/src/instanceOf.ts

```diff
@@ -0,0 +1,56 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+/**
+ * Attempt to pin "load-time" implementations of call and toString to avoid
+ * behavior changes at "run-time" caused by unexpected prototype manipulation
+ */
+const pinnedObjectToString: (value: any) => string = Function.prototype.call.bind(
+    Object.prototype.toString, // eslint-disable-line @typescript-eslint/unbound-method
+);
+
+const toString: (value: any) => string = (value) => pinnedObjectToString(value);
+
+const objectString = toString(Object.prototype);
+
+/**
+ * Plain-Old-JavaScript-Object -- as recommended by TypeScript
+ */
+export type POJO = Record<string, unknown>;
+
+/**
+ * Tests if the passed value is a plain-old-javascript-object
+ * @param value - the value to check
+ * @returns true if the value is an object, otherwise false
+ */
+export function instanceOfObject(value: any): value is POJO {
+    return value !== null && typeof value === "object";
+}
+
+/**
+ * Produces partially applied functions that test if passed values match the provided builtin instance type
+ * @param match - an instance of the builtin to match subsequent values with
+ * @returns the partially applied test function -- (value: any): value is typeof match
+ */
+export function bindInstanceOfBuiltin<T>(match: T) {
+    const compareString = toString(match);
+
+    if (compareString === objectString) {
+        throw new Error(`bindInstanceOfBuiltin cannot classify '${objectString}' instances`);
+    }
+
+    return (value: any): value is T => compareString === toString(value);
+}
+
+export const instanceOf$ = Object.freeze({
+    Object: instanceOfObject,
+
+    ArrayBuffer: bindInstanceOfBuiltin(ArrayBuffer.prototype),
+    Error: bindInstanceOfBuiltin(new Error("N/A")),
+    Map: bindInstanceOfBuiltin(Map.prototype),
+    Set: bindInstanceOfBuiltin(Set.prototype),
```

Comment:
We should minimize the case we need to check the type of object.

Looks like the only usage of instanceof Map/Set is in property-inspector-table.  @DLehenbauer , is there anyway to avoid it?

Reply:
@curtisman , for the purposes of this PR, I can remove the `Map` and `Set` checkers from the export (and let property-inspector-table leverage `bindInstanceOfBuiltin` if not some other approach -- i.e. defer decisions on how to refactor it) if that is preferable.

Reply:
I wonder whether we can do away with Error as well.  😄 

Most of the usage are node only usage in command line tools or server code, so using instanceof probably isn't an issue.
But for consistency sake, we probably want to get rid usage of instanceof and have a lint rule enforcing it. 

Most cases, it might just need to check if the property `stack` is there and that would be good enough check.